### DISCUSSION TITLE
Welcome to BLEUnlock Discussions!

### DISCUSSION THREAD

**Discussion by ts1** [ts1/BLEUnlock#89] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by mansidak** [comment]
Love your work with BLEUnlock, man! Just got you a coffee :) 

Did have a quick question that I figured you'd be able to answer. I was working on a Swift mac project that required me to run some sudo terminal commands and I as curious how you implemented the functionality where you request user login password and leverage that to unlock that mac? I couldn't pin point the exact portion in your code that I could gather inspiration from. Was wondering if you had any two cents on how to achieve this functionality of running a sudo command which requires user login password input. Would it be Authorization services? or keychain? Lemme know if you have any thoughts! Have a great rest of your week :)

**Reply by ts1** [reply]
Thanks for the coffee!

BLEUnlock asks for the password in a simple `NSAlert` [here](https://github.com/ts1/BLEUnlock/blob/a8edb92cbc97d2fbd495823a29aae1ec63af1c7d/BLEUnlock/AppDelegate.swift#L445).
Then it stores the password in the Keychain.

On the lock screen it enters the password by faking keystrokes [here](https://github.com/ts1/BLEUnlock/blob/a8edb92cbc97d2fbd495823a29aae1ec63af1c7d/BLEUnlock/AppDelegate.swift#L247).

I know the above is not what you want to do, running the command in sudo is another story.
I think you need to gain privileges somehow, but I have never done it before, I have no idea how to achieve that.
Sorry I couldn't help.

**Reply by mansidak** [reply]
@ts1 Gotchu man :) 

And thanks for the answer. I do have one question though: how do you verify that the password entered was correct? Are you just asking for password and saving it in keychain and then retreiving it without somehow checking that it's the correct password? 

The reason I ask is because in system preferences BLEUnlock comes under the apps that can make changes so I'm trying to figure out which process did you use to get admin privileges? Because if I can do that like you did, I can then run sudo command.

**Reply by ts1** [reply]
It doesn't verify that the password is correct.
If a user enters an incorrect password, he or she will simply see the lock screen with the incorrect password.

Accessibility permission is obtained [here](https://github.com/ts1/BLEUnlock/blob/a8edb92cbc97d2fbd495823a29aae1ec63af1c7d/BLEUnlock/AppDelegate.swift#L664).
Bluetooth privilege is probably [here](https://github.com/ts1/BLEUnlock/blob/a8edb92cbc97d2fbd495823a29aae1ec63af1c7d/BLEUnlock/Info.plist#L27).

I don't get admin privileges, but the system does.

**Reply by mansidak** [reply]
Gotchu. One last question, does your program retrieve the password from the keychain every time the mac is unlocked? If so, why doesn't it ask for a user password every time it retrieves the password? Because every time I try to run the command to retrieve anything from the keychain, the system asks for password, butin BLEUnlock's case it doesn't. :/

**Reply by ts1** [reply]
I'm not very sure, but this might be related to code signing.
When I run the debug version of BLEUnlock which is not signed by Apple developer signature, it asks for password every time.

**Comment by Ipatmegit** [comment]
Hi, how do I remove the BLEUnlock installed using the zip download?
